### ISSUE TITLE
Allow arbitrary serialize/deserialize mappings

### ISSUE THREAD

**Issue by nornagon** [json4s/json4s#79]
Currently, it's really hard to write a system of serializers and deserializers for the following sort of situation:

``` scala
trait Tree
case class Node(val left: Tree, val right: Tree) extends Tree
case class Leaf(val value: Int) extends Tree
```

`TypeHints` could be used, but as written, there can only be a single `typeHintFieldName` for an entire `Formats`. I would like to use "name" for some type hints, but "id" for others -- and really, I just want to be able to write a custom serializer that can 'shell out' to the default serialization: something like, "to serialize a Tree, do whatever you would normally do then add {"type":c.getClass.getSimpleName}".

**Comment by slivkamiro** [comment]
Is there any progress on this? I would like multiple type hint field names too.

**Comment by magnolia-k** [comment]
@slivkamiro 
If the patch is complete, can you give me a PR?

**Comment by slivkamiro** [comment]
Hi, I have some failing tests I need to resolve first. It'll take a while, but I will open a PR when I'm done.

**Comment by slivkamiro** [comment]
Any volunteers for a review?

**Comment by magnolia-k** [comment]
@slivkamiro 
I'll review it next weekend

**Comment by mccartney** [comment]
Fixed in #656
